algorithms/kernel/dtrees/forest/classification/oneapi/cl_kernels/df_batch_classification_kernels.cl

```diff
@@ -0,0 +1,573 @@
+
+        const int rowsOffset = nodeList[nodeId * nNodeProp + 0];
+        const int nRows      = nodeList[nodeId * nNodeProp + 1];
+
+        // each sub group will process 16 bins and produce 1 best split for it
+        // expect maximum 16 subgroups only for each node
+        const int maxBinsBlocks = 16;
+        __local algorithmFPType bufI[maxBinsBlocks]; // storage for impurity decrease
+        __local int bufS[maxBinsBlocks * nNodeProp]; // storage for split info
+
+        const algorithmFPType minImpDec = (algorithmFPType)-1e30;
+        const int valNotFound           = 1 << 30;
+
+        algorithmFPType curImpDec = minImpDec;
+        int curFeatureValue       = leafMark;
+        int curFeatureId          = leafMark;
+
+        nodeList[nodeId * nNodeProp + 2] = curFeatureId;
+        nodeList[nodeId * nNodeProp + 3] = curFeatureValue;
+        nodeList[nodeId * nNodeProp + 4] = nRows;
+
+        algorithmFPType mrgN = (algorithmFPType)nRows;
+
+        algorithmFPType bestLN = (algorithmFPType)0;
+
+        algorithmFPType mrgCls[nProp] = { (algorithmFPType)0 };
+
+        algorithmFPType imp = (algorithmFPType)1;
+
+        int totalBins = 0;
+
+        // totalBins is calculated by each subgroup
+        for (int featIdx = sub_group_local_id; featIdx < nSelectedFeatures; featIdx += sub_group_size)
+        {
+            int featId = selectedFeatures[nodeId * nSelectedFeatures + featIdx];
+            int nBins  = binOffsets[featId + 1] - binOffsets[featId];
+            totalBins += sub_group_reduce_add(nBins);
+        }
+        totalBins = sub_group_broadcast(totalBins, 0);
+
+        int currFtrIdx  = 0;
+        int featId      = selectedFeatures[nodeId * nSelectedFeatures + currFtrIdx];
+        int binId       = 0;
+        int currFtrBins = binOffsets[featId + 1] - binOffsets[featId];
+        int passedBins  = 0;
+
+        for (int i = local_id; i < totalBins; i += local_size)
+        {
+            while (i >= passedBins + currFtrBins)
+            {
+                passedBins += currFtrBins;
+                currFtrIdx++;
+                featId      = selectedFeatures[nodeId * nSelectedFeatures + currFtrIdx];
+                currFtrBins = binOffsets[featId + 1] - binOffsets[featId];
+            }
+            binId = i - passedBins;
+
+            algorithmFPType mrgLRN[2]           = { (algorithmFPType)0 };
+            algorithmFPType mrgLRCls[nProp * 2] = { (algorithmFPType)0 };
+
+            // calculating classes histogram
+            for (int row = 0; row < nRows; row++)
+            {
+                int id      = treeOrder[rowsOffset + row];
+                int bin     = data[id * nFeatures + featId];
+                int classId = (int)response[id];
+                mrgLRN[(int)(bin > binId)] += (algorithmFPType)1;
+                mrgLRCls[nProp * (int)(bin > binId) + classId] += (algorithmFPType)1;
+            }
+
+            imp                  = (algorithmFPType)1;
```

Comment:
In such cases, it make sense to create couple of auxiliary variables like:
```
const algorithmFPType zero = 0.0;
const algorithmFPType one = 1.0;
```
This will make code more readable, and will help to avoid typos.

Reply:
Ok, thank you, will update it on next iteration.